fastn-core/src/migrations.rs

```diff
@@ -0,0 +1,210 @@
+const MIGRATION_TABLE: &str = r#"
+
+CREATE TABLE IF NOT EXISTS
+    fastn_migration
+(
+    id               INTEGER PRIMARY KEY,
+    app_name         TEXT NOT NULL,
+    migration_number INTEGER NOT NULL UNIQUE,
+    migration_name   TEXT NOT NULL,
+    applied_on       INTEGER NOT NULL
+) STRICT;
+
+"#;
+
+pub(crate) async fn migrate(config: &fastn_core::Config) -> Result<(), MigrationError> {
+    let db = config.get_db_url().await;
+    create_migration_table(config, db.as_str()).await?;
+
+    if !has_new_migration(config, db.as_str()).await? {
+        return Ok(());
+    }
+
+    let migrations_to_apply = find_migrations_to_apply(config, db.as_str()).await?;
+
+    let now = chrono::Utc::now();
+    for migration in migrations_to_apply {
+        config
+            .ds
+            .sql_batch(db.as_str(), migration.content.as_str())
+            .await?;
+        mark_migration_applied(&config, db.as_str(), &migration, &now).await?;
+    }
+
+    Ok(())
+}
+
+async fn create_migration_table(
+    config: &fastn_core::Config,
+    db: &str,
+) -> Result<(), fastn_utils::SqlError> {
+    config.ds.sql_batch(db, MIGRATION_TABLE).await?;
+    Ok(())
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum MigrationError {
+    #[error("Sql Error: {0}")]
+    SqlError(#[from] fastn_utils::SqlError),
+    #[error("Cannot delete applied migration")]
+    AppliedMigrationDeletion,
+    #[error("The migration order has changed or its content has been altered")]
+    AppliedMigrationMismatch,
+}
+
+async fn has_new_migration(config: &fastn_core::Config, db: &str) -> Result<bool, MigrationError> {
+    let last_available_migration = match config.package.migration.last() {
+        Some(last_available_migration) => last_available_migration,
+        None => return Ok(false),
+    };
+
+    let results = config
+        .ds
+        .sql_query(
+            db,
+            format!(
+                r#"
+                    SELECT
+                        migration_number
+                    FROM
+                        fastn_migration
+                    WHERE
+                        app_name = '{}'
+                    ORDER BY migration_number DESC
+                    LIMIT 1;
+                "#,
+                config.package.name
+            )
+            .as_str(),
+            vec![],
+        )
+        .await?;
+
+    match results.len() {
+        0 => Ok(true),
```

Comment:
Overall feature may work fine, but this statement is wrong. If there are no migrations in the db it does not on its own mean we have unapplied migrations. 

Also in general if some task is done once in one way, unless we have strong reason, we should do it the next time in same way. The implementation of migration in ft-sdk is differnt from there. So unless you have found fault in that implementation (eg if you feel this implementation is technically better) I prefer we use the same Algo as there. Same function names etc, same behaviour etc.

Reply:
done